Editor toggling, visual design spec. integration, various stylistic tweaks

sonicyeti opened the pull request:
This PR covers the incorporation of accordion style functionality for unit tests including: 

- Unit tests can now be toggled to expand or collapse
- A single unit test's max height caps at around the 20 line mark, where scrolling takes over to preserve real estate
- The unit test header has been brought closer to alignment with the original design specification
- Unit Test is now a list group item type from a component standpoint
- Swapping play and delete iconography positioning
- Resolution for multiple Code Mirror instances introducing lag

**Screenshot:**
![Screen Shot 2020-09-10 at 10 00 43 AM](https://user-images.githubusercontent.com/52717970/92743135-4773ee00-f34e-11ea-8448-3437f9660f02.png)

**Demo:**
![2020-09-10 10 19 42](https://user-images.githubusercontent.com/52717970/92744796-cae20f00-f34f-11ea-9755-d60a603003c3.gif)

Closes #2562
Fixes INS-85
Fixes INS-48
Fixes INS-52

sonicyeti committed:
Editor toggling, visual design spec. integration, various stylistic tweaks

develohpanda commented on packages/insomnia-app/app/ui/components/unit-test-selection.js:
```diff
@@ -0,0 +1,29 @@
+type Props = {
+  onChange: Function,
+  unitTest: Object,
+  selectableRequests: Object,
```
It would be great if we could type these more explicitly.

For example, just from looking at the usage in the component
```suggestion
  onChange: () => void,
  selectedRequestId?: string,
  selectableRequests: Array<{name: string, request: { _id: string }}>,
```

develohpanda commented on packages/insomnia-app/app/ui/components/unit-test-selection.js:
```diff
@@ -0,0 +1,29 @@
+  selectableRequests: Object,
+};
+
+export default function UnitTestSelection({ onChange, unitTest, selectableRequests }: Props) {
```
Suggestion to rename component as `UnitTestRequestSelector`, as this is specific to selecting requests 😄

develohpanda commented on packages/insomnia-app/app/ui/components/wrapper-unit-test.js:
```diff
@@ -351,45 +352,28 @@ class WrapperUnitTest extends React.PureComponent<Props, State> {
+            unitTest={unitTest}
+            selectableRequests={selectableRequests}
+          />
+        }
```
I like that you extracted `UnitTestItem` and `UnitTestSelection` out into standalone components 👍 

We can avoid having to pass `<UnitTestSelection />` as a prop into `UnitTestItem`; instead, we can simply pass the onChange handler as a prop, for instance, `onRequestSelectionChanged={this._handleSetActiveRequest.bind(this, unitTest)}`.

Then, `UnitTestItem` becomes

```js
function UnitTestItem(props) {
  ...
  <UnitTestSelection onChange={props.onRequestSelectionChanged} ... />
  ...
}
```
instead of
```js
function UnitTestItem(props) {
  ...
  {props.testSelection}
  ...
}
```

This way UnitTestSelection can be moved into insomnia-components!

develohpanda commented on packages/insomnia-app/app/ui/components/wrapper-unit-test.js:
```diff
@@ -351,45 +352,28 @@ class WrapperUnitTest extends React.PureComponent<Props, State> {
+            selectableRequests={selectableRequests}
+          />
+        }
+        onDeleteTest={() => this._handleDeleteTest(unitTest)}
+        onRunTest={() => this._handleRunTest(unitTest)}
```
Passing arrow functions as props mean a new function is instantiated on each render. A newly instantiated function, even if it has the same implementation, is unique. Because it is unique, React will tell `UnitTestItem` to update, because a prop has changed, when in-fact it hasn't. Try use the `.bind` syntax within class components, or `React.useCallback` within function components 😄

develohpanda commented on packages/insomnia-components/components/list-group/list-group.stories.js:
```diff
@@ -47,10 +48,47 @@ const tests = [
+  <div style={{ width: '100%' }}>
+    <ListGroup>
+      {unitTests.map((test, i) => (
+        <UnitTestItem item={test} key={i}>
```
Index as keys hehe. Doesn't really matter here, but good to get into the practice of avoiding it. I suggest to use `test._id` as the key

develohpanda commented on packages/insomnia-app/app/ui/components/wrapper-unit-test.js:
```diff
@@ -351,45 +352,28 @@ class WrapperUnitTest extends React.PureComponent<Props, State> {
+        onDeleteTest={() => this._handleDeleteTest(unitTest)}
+        onRunTest={() => this._handleRunTest(unitTest)}
+        testsRunning={testsRunning}
+        testNameEditable={
+          <UnitTestEditable
+            onSubmit={this._handleChangeTestName.bind(this, unitTest)}
+            value={unitTest.name}
+          />
+        }>
```
Same as above, but less trivial because `Editable` isn't in `insomnia-components`. _But_, it would be a great idea to pull `Editable` into the component library, and other usages can be updated later on.

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-item.js:
```diff
@@ -0,0 +1,108 @@
+  testNameEditable,
+  testsRunning,
+}: Props) => {
+  const [collapse, toggleCollapse] = React.useState(false);
```
We can use the `useToggle` hook from `react-use`, like in the new sidebar hehe

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-item.js:
```diff
@@ -0,0 +1,108 @@
+          style={!collapse ? { transform: `rotate(${toggleRotation}deg)` } : {}}>
+          <SvgIcon icon="chevron-down" />
+        </Button>
+        <Button variant="text" disabled>
+          <SvgIcon icon="file" />
+        </Button>
```
Does this icon need to exist inside a disabled `Button`? I'm guessing it's for consistent spacing

sonicyeti replied:
Exactly. Disabled happened to match the design spec while reducing custom styling overhead.

develohpanda reviewed (changes requested):
Nicely done 😄 

I've made a couple of comments, but one overarching one is to do with defining Props for components. Because props are the interface between a component and it's consumer, we should try to be as explicit as possible with the prop type definitions, rather than generic `Function`, `Object`, etc. It will only help us when we come to look at or consume the same component in a few months time.

I will add that there are many components currently that do not define their prop types clearly; those components were likely created before a typing methodology was introduced. Even though we are currently using Flow right now, when we eventually look to move towards TS, the flow type definitions will be useful. 😄

sonicyeti committed:
PR Feedback

Prop type specificity, component naming, proper binding via useCallback,  useToggle

DMarby commented on packages/insomnia-app/app/ui/components/wrapper-unit-test.js:
```diff
@@ -354,20 +353,18 @@ class WrapperUnitTest extends React.PureComponent<Props, State> {
     const { testsRunning } = this.state;
     const selectableRequests = this.buildSelectableRequests();
 
+    console.log(selectableRequests);
```
I think this might be a leftover debug statement that should be removed?

sonicyeti replied:
Thanks for the catch, pulling it now!

sonicyeti committed:
Pulling log

develohpanda commented on packages/insomnia-components/components/list-group/list-group.stories.js:
```diff
@@ -73,12 +77,119 @@ const unitTests = [
+          onChange={_handleRequestSelected}
+          selectableRequests={unitTestRequests}>
+          <div style={{ padding: '10px 15px', border: '1px solid #eee', margin: '10px' }}>
+            CodeMirror instance.
```
👌

develohpanda commented on packages/insomnia-components/components/list-group/list-group.stories.js:
```diff
@@ -89,7 +200,7 @@ export const _unitTestResults = () => (
     <ListGroup>
       {unitTestResults.map((test, i) => (
-        <UnitTestResultItem item={test} key={i} />
+        <UnitTestResultItem item={test} key={test._id} />
```
👌

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-item.js:
```diff
@@ -63,27 +49,36 @@ const UnitTestItem = ({
+  const toggleIconRotation = -90;
+  const _handleItemToggle = React.useCallback(() => {
+    toggle();
+  }, [toggle]);
```
I think this can be removed in favor of just using `onClick={toggle}`

The `React.useCallback` would be useful if you were to do `onClick={() => { toggle() }}`, because the `() => {}` syntax creates a new function. 🙂

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-item.js:
```diff
@@ -1,41 +1,27 @@
+  onRunTest?: () => void,
   testNameEditable?: React.Node,
   testsRunning?: Object,
+  onChange: () => void,
```
It's not immediately clear what this change handler is for - changing the test body, title, request, etc?

Similar to the other event handles (`onDeleteTest`, `onRunTest`), I suggest naming this explicitly like `onSetActiveRequest`, for example.

`onChange` on it's own is fine for a generic component like `Editable`, though. Contextual 😄

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-item.js:
```diff
@@ -1,41 +1,27 @@
-  onDeleteTest?: Function,
-  onRunTest?: Function,
-  testSelection?: React.Node,
+  onDeleteTest?: () => void,
+  onRunTest?: () => void,
```
I'd like to double check but I don't think these callbacks should be optional, right?

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-item.js:
```diff
@@ -1,41 +1,27 @@
 const StyledResultListItem: React.ComponentType<{}> = styled(ListGroupItem)`
   padding: 0 var(--padding-sm);
 
-  .form-control {
```
Yay, happy to see this move into what uses it specifically. 😄

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-item.js:
```diff
@@ -94,13 +89,7 @@ const UnitTestItem = ({
-      {collapse && (
-        <div
-          initial={{ height: collapse ? '100%' : '0px' }}
-          animate={{ height: collapse ? '100%' : '0px' }}>
```
Was the removal of animation intentional?

sonicyeti replied:
~~I pulled this because while we can slide-reveal the CM instance, when collapsed, it no longer exists in the DOM and looks quite janky when attempting to slide-hide and wanted to avoid having a toggle collapsed callback tied to the tween duration just for the sake of animating the exit of the CM instance.~~

Decided to keep and enable it. It just looks so much better sliding into view despite it immediately disappearing on collapse.

develohpanda replied:
Fair enough, we can alter it depending on feedback. Either way seems fine

develohpanda commented on packages/insomnia-components/index.js:
```diff
@@ -33,6 +34,7 @@ export const ListGroup = _ListGroup;
 export const ListGroupItem = _ListGroupItem;
 export const UnitTestItem = _UnitTestItem;
 export const UnitTestResultItem = _UnitTestResultItem;
+export const UnitTestRequestSelector = _UnitTestRequestSelector;
```
We can get away with not exporting this selector from insomnia-components since it's more of an internal component. Keen for your thoughts on this however - what stuff is exported from the library and what is not.

sonicyeti replied:
100% agreed, I didn't realize that this export serves the sole purpose of exposing it for usage in app. I thought it was necessary for bundling deps which the UT result item needs. I'll yank it!

develohpanda reviewed (approved):
A couple of minor notes, good to merge after they are addressed. 👍

sonicyeti committed:
PR Feedback

Prop name specificity, prevent export of one-off request selector, set defaultValue on UT selector, clean up legacy animation props.

sonicyeti committed:
Merge branch 'develop' into unit-testing-cm-accordion

sonicyeti committed:
Fixing unit test item component story

sonicyeti committed:
Fixing UT content slide reveal
